### DISCUSSION TITLE
Unity Plugin: Accessing contact and collision data

### DISCUSSION THREAD

**Discussion by makeplayhappy** [google-deepmind/mujoco#144] [Asking for Help] [answered]
Are there any available collision callbacks?
I'm wondering what the best way to be informed about when collisions are happening. 
Currently I'm doing this:
```
if( Data != null && Data->contact != null && Data->ncon > 0 ){  
      MujocoLib.mjContact_* contact = Data->contact;  
      for(int i = 0; i < Data->ncon; i++){  
        // match up contact->geom1 contact->geom2 with _orderedComponents[].MujocoId and trigger a collision callback / event
        contact++;
      }  
 }  
```
I imagine I'm missing something but I can't find anything in the code or docs that suggest any other way.

**Comment by erez-tom** [comment, marked as answer]
If you want to know when whether a specific geom is touching anything, you can create a site that envelopes it, assign a [touch sensor](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-touch) to the site, and read the [sensor value](https://github.com/deepmind/mujoco/blob/main/unity/Runtime/Components/MjSiteScalarSensor.cs#L32) (which, in this case, will be the normal force).

If you want the list of geoms in contact with a certain geom, you would indeed have to scan the list of `Data->contact` to find that out.

Invoking collision detection manually is possible in principle ([this](https://github.com/deepmind/mujoco/blob/main/include/mujoco.h#L68) is where the specific collision callbacks reside, and [here](https://mujoco.readthedocs.io/en/latest/programming.html#contacts) there's some explanation how it works).  Particularly, if you want to know whether a specific pair of geoms is in contact, the MuJoCo Way would be to define a [`pair`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#contact-pair) and force collision detection on this pair at every timestep.  However:
- we haven't yet created a component for `pair`.
- You'd still have to scan the list of contacts to find out their relative distance -- the `pair` object won't give you direct access to the relevant contacts (an interesting feature request -- I will add it to our list).

See also #123.

Does this explanation cover your intended use-case?

**Reply by makeplayhappy** [reply]
Ah, touch sensor is what I was looking for and should be enough for my use case, thanks.
